bin/cylc

```diff
@@ -14,643 +15,225 @@
+            PATH_STR="${PATH_HEAD}:${PATH_TAIL}"
+        done
+        echo "${ITEM_STR}:${PATH_STR}"
+    else
+        echo "${PATH_STR}"
+    fi
+}
+
+abort_bad_command() {
+    echo "${CYLC_NS}: ${UTIL}: unknown utility. Abort." >&2
+    echo "Type \"${CYLC_NS} help all\" for a list of utilities." >&2
+    exit 1
+}
+
+abort_ambiguous_command() {
+    echo "\"${UTIL}\" is too ambiguous to be a valid Cylc command abbreviation."
+    echo "These commands match the abbreviation \"${UTIL}\":"
+    ls "${CYLC_HOME_BIN}/cylc-${UTIL}"* 2>/dev/null \
+        | sed -r "s,${CYLC_HOME_BIN}/cylc-,,"
+    exit 1
+}
+
+run_cylc_command_matched() {
+    COMMAND=$(ls "${CYLC_HOME_BIN}/cylc-${UTIL}"*)
+    if [[ ! -f "${COMMAND}" || ! -x "${COMMAND}" ]]; then
+        abort_bad_command
+    fi
+    CYLC_UTIL="$(echo "${COMMAND}" | sed -r "s,${CYLC_HOME_BIN}/cylc-,,")"
+    export CYLC_UTIL
+    exec "${COMMAND}" "$@"
+    exit 0
+}
+
+run_cylc_command() {
+    COMMAND=$(ls "${CYLC_HOME_BIN}/cylc-${UTIL}")
+    if [[ ! -f "${COMMAND}" || ! -x "${COMMAND}" ]]; then
+        abort_bad_command
+    fi
+    CYLC_UTIL="${UTIL}"
+    export CYLC_UTIL
+    exec "${COMMAND}" "$@"
+    exit 0
+}
+
+init_cylc
+
+UTIL="help"
+if (( "$#" > 0 )); then
+    UTIL="$1"
+    shift 1
+fi
+# Test if the user has provided a valid cylc command, if so, run it.
+if [[ -f "${CYLC_HOME_BIN}/cylc-${UTIL}" && "${UTIL}" != 'help'  ]]; then
+    run_cylc_command "$@" 
+    exit 0
+fi
+
+case "${UTIL}" in
+# Deal with categories
+control|information|all|task|license|GPL|admin|preparation|hook|discovery\
+        |utility|util|prep)
+    if (( "$#" == 0 )); then
+        help_util "${UTIL}"
+        exit 0
+    elif (( "$#" > 0 )); then
+        UTIL="$1"
+        # Discard the category qualifier argument by shifting
+        # the arguments along by 1.
+        shift 1
+    fi
+    :;;
+help|h|?|--help|-h)
+    help_util "$@"
+    exit 0
+    :;;
+version|--version|-V)
+    print_version "$@"
+    exit 0
+    :;;
+categories)
+    category_help_command "$@"
+    exit 0
+    :;;
+esac
+
+# User has not given an exact cylc command, or a help, category, or version
+# command.
+# So now deal with matching commands and disambiguation.
+
+# Abort if no matches to command abbrev, list matches if multiple matches
+test "$(ls "${CYLC_HOME_BIN}/cylc-${UTIL}"* 2>/dev/null)" \
+    || { abort_bad_command; exit 1; }
+# Check if command (singular) exists first.
+# Needed for handling "graph" and "graph-diff"
+test "$(ls "${CYLC_HOME_BIN}/cylc-${UTIL}" 2>/dev/null | wc -l )" \
+    -eq 1 && { run_cylc_command "$@"; exit 1; }
+# Now try and match abbreviated commands.
+test "$(ls "${CYLC_HOME_BIN}/cylc-${UTIL}"* 2>/dev/null | wc -l )" \
+    -eq 1 && { run_cylc_command_matched "$@"; exit 0; }
+# Now try and match abbreviated commands.
+test "$(ls "${CYLC_HOME_BIN}/cylc-${UTIL}"* 2>/dev/null | wc -l )" \
+    -gt 1 && { abort_ambiguous_command; exit 1; }
```

Comment:
A few issues here:

1. We should reduce the number of `ls`, where possible, to minimise interaction with the file system.
2. I think it is neatly to use `if ...; then ...; elif ...; then ... else ...; fi` block here.
3. The `exit` statement at the end of each block appears to be unnecessary.